### ISSUE TITLE
Use a single step setup for C++-API and all bindings

### ISSUE THREAD

**Issue by BenjaminRueth** [precice/precice#366]
Currently, some bindings ([C++](https://github.com/precice/precice/wiki/Adapter-Example), [Python](https://github.com/precice/precice/blob/e480aef4e3a74ccd405f03a2770bee18cdb548e9/src/precice/bindings/python/precice.pyx#L112-L122)) use a two step setup, where first the SolverInterface object is created and then the SolverInterface is configured:

```
SolverInterface( String participantName, int rank, int size );
configure( String configuration );
```

Other bindings ([C](https://github.com/precice/precice/blob/e480aef4e3a74ccd405f03a2770bee18cdb548e9/src/precice/bindings/c/SolverInterfaceC.h#L18-L22), Fortran) use a single step setup:

```
precicec_createSolverInterface(participantName, configuration, rand, size);
```

This is not consistent and (at least for me) there is no obvious reason why a two-step setup is necessary. Therefore, let's change all bindings to use a one step setup.

**Note:** If we want to have this change in `v1.x`, we have to maintain the API and therefore have to provide the two step setup. Only in `v2.0` we can entirely get rid of the two-step setup.

**Comment by uekerman** [comment]
I'll remove the `good first issue` label as this has implications everywhere (all integration tests, all adapters, all bindings).

**Comment by BenjaminRueth** [comment]
https://github.com/precice/python-bindings/pull/13 implements this change for the python bindings (which is good). However, https://github.com/precice/python-bindings/pull/13 currently relies on the two-level C++ interface, because the C++ interface does not provide the single step setup yet.

**Comment by uekerman** [comment]
I'll suggest to add an additional constructor 
```c++
SolverInterface( String participantName, String configurationFileName, int rank, int size );
```
to the C++ API, which calls the old constructor and `configure`, and add a deprecated warning to the old constructor (not sure yet how to do this technically). The documentation and all official adapters use the new variant.

Why:
* For backwards compatibility. Breaking the API is much more severe than breaking the configuration in my opinion. We could then remove the old variant with v3.0. I would not want to make it a fixed law to do it this way every time. This time, it is easy to do and critical at the same time.
* Because we might need the split into constructor and `configure` in our integration tests. I suspect that this was also the reason for the design in the first place.

**Comment by fsimonis** [comment]
@uekerman I agree with your suggested constructor.

However, we should definitely break the API now with the upcoming version 2.
The configuration is integral part of the construction of SolverInterface and both functions belong together. There is no valid function that can be called between the constructor and the configuration. A single-step setup will lead to less confusion among users, reduces the required documentation and reduced the state of the interface.

Our `C` and `Fortran` bindings already use the single-step setup and the change for the python bindings is in progress https://github.com/precice/python-bindings/pull/13. Changing this in the `C++` interface as well is just the next step on the path to a better API.

This change will break user-code, but in a trivial way. Constructing and configuring has to happen in sequence anyhow. Thus, this change requires to merge two following lines:
https://github.com/precice/precice/blob/30b2a2d17c1d8586e7ce0e8801e73a4b4a98f0c7/tools/solverdummies/cpp/solverdummy.cpp#L32-L33
```diff
- SolverInterface interface(solverName, commRank, commSize);
- interface.configure(configFileName);
+ SolverInterface interface(solverName, configFileName, commRank, commSize);
```

Deprecating the configuration will result in 4 constructors and a configuration function, which introduces additional complexity in the state graph as well as the required documentation and make user support more difficult: What happens if a user constructs the `SolverInterface` with the new method, but then calls configure? How do we document this?

**TLDR**
The only sane way to go forward is to unify the setup steps as it:
* prevents one more pitfall for the user
* leaves us with 2 constructors
* simplifies the documentation
* simplifies the state graph by removing one node
* makes the core library easier to maintain

The downside is that:
* the user has to change 2 lines
* we have to adjust our tests (on it)

**Comment by MakisH** [comment]
@fsimonis suggestion sounds reasonable. However, we should port all the adapters to preCICE v2 and pull the cable from any support for preCICE v1, since we don't have the resources to also support different preCICE versions, alongside the different solver versions.

I think the main difference between what @uekerman and what @fsimonis suggest is a different answer to the question: _"should we allow people that are already using preCICE v1 (e.g. because they already have it installed in their shared cluster/system) to still get recent versions of the adapters?"_. Although I like the _"this is deprecated - upgrade at your own pace"_ suggestion, the upgrading may not be so difficult at the end of the day.

**Comment by uekerman** [comment]
Then OK for me. 

> What happens if a user constructs the SolverInterface with the new method, but then calls configure? How do we document this?

... is indeed a point I didn't realize. Let's break the API for good then.

**Comment by BenjaminRueth** [comment]
python and MATLAB bindings are almost done: https://github.com/precice/python-bindings/pull/18 and https://github.com/precice/matlab-bindings/pull/9

**Comment by BenjaminRueth** [comment]
All bindings now use the single step setup. This closes the issue.
